orca-kotlin/src/main/kotlin/com/netflix/spinnaker/orca/ext/Stage.kt

```diff
@@ -83,6 +83,12 @@ fun Stage.anyUpstreamStagesFailed(): Boolean =
 fun Stage.syntheticStages(): List<Stage> =
   execution.stages.filter { it.parentStageId == id }
 
+fun Stage.recursiveSyntheticStages(): List<Stage> =
+  syntheticStages() + syntheticStages().flatMap {
+    if (it.syntheticStages().isEmpty()) emptyList()
```

Comment:
Don't think you need this `if`, if the stages are empty, the recursive `it.recursiveSyntheticStages()` never makes it into the `flatmap`, since there are no synthetic stages.